util/retry/retry.go

```diff
@@ -18,107 +18,93 @@
-type MaxAttemptsError struct {
-	MaxAttempts int
+	r := Retry{opts: opts}
+	r.Reset()
+	return r
 }
 
-// Error implements error interface.
-func (re *MaxAttemptsError) Error() string {
-	return fmt.Sprintf("maximum number of attempts exceeded %d", re.MaxAttempts)
+// Reset resets the Retry to its initial state, meaning that the next call to
+// Next will return true immediately and subsequent calls will behave as if
+// they had followed the very first attempt (i.e. their backoffs will be
+// short).
+func (r *Retry) Reset() {
+	r.currentAttempt = 0
+	r.isReset = true
 }
 
-const (
-	// Break indicates the retry loop is finished and should return
-	// the result of the retry worker function.
-	Break Status = iota
-	// Reset indicates that the retry loop should be reset with
-	// no backoff for an immediate retry.
-	Reset
-	// Continue indicates that the retry loop should continue with
-	// another iteration of backoff / retry.
-	Continue
-)
+func (r Retry) retryIn() time.Duration {
+	backoff := float64(r.opts.InitialBackoff) * math.Pow(r.opts.Multiplier, float64(r.currentAttempt))
+	if maxBackoff := float64(r.opts.MaxBackoff); backoff > maxBackoff {
+		backoff = maxBackoff
+	}
 
-// Options provides control of retry loop logic via the
-// WithBackoffOptions method.
-type Options struct {
-	Tag         string        // Tag for helpful logging of backoffs
-	Backoff     time.Duration // Default retry backoff interval
-	MaxBackoff  time.Duration // Maximum retry backoff interval
-	Constant    float64       // Default backoff constant
-	MaxAttempts int           // Maximum number of attempts (0 for infinite)
-	UseV1Info   bool          // Use verbose V(1) level for log messages
-	Stopper     *util.Stopper // Optionally end retry loop on stopper signal
+	var delta = r.opts.RandomizationFactor * backoff
+	// Get a random value from the range [backoff - delta, backoff + delta].
+	// The formula used below has a +1 because time.Duration is an int64, and the
+	// conversion floors the float64.
+	return time.Duration(backoff - delta + rand.Float64()*(2*delta+1))
 }
 
-// WithBackoff implements retry with exponential backoff using
-// the supplied options as parameters. When fn returns Continue
-// and the number of retry attempts haven't been exhausted, fn is
-// retried. When fn returns Break, retry ends. As a special case,
-// if fn returns Reset, the backoff and retry count are reset to
-// starting values and the next retry occurs immediately. Returns an
-// error if the maximum number of retries is exceeded or if the fn
-// returns an error.
-func WithBackoff(opts Options, fn func() (Status, error)) error {
-	backoff := opts.Backoff
-	tag := opts.Tag
-	if tag == "" {
-		tag = "invocation"
+// Next returns whether the retry loop should continue, and blocks for the
+// appropriate length of time before yielding back to the caller. If a stopper
+// is present, Next will eagerly return false when the stopper is stopped.
+func (r *Retry) Next() bool {
```

Comment:
This is a strangely named method. `Next()` seems like it wouldn't return anything. `Valid() bool` seems better and matches how we use iterators elsewhere.